src/payment/strategies/stripe-upe/stripe-upe-payment-strategy.ts

```diff
@@ -95,55 +96,63 @@ export default class StripeUPEPaymentStrategy implements PaymentStrategy {
             await this._store.dispatch(this._storeCreditActionCreator.applyStoreCredit(useStoreCredit));
         }
 
-        try {
-            if (includes(APM_REDIRECT, methodId)) {
-                await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
-                const paymentMethod = this._store.getState().paymentMethods.getPaymentMethodOrThrow(methodId);
-                const paymentPayload = {
-                    methodId,
-                    paymentData: {
-                        formattedPayload: {
-                            credit_card_token: { token: paymentMethod.clientToken },
-                            vault_payment_instrument: shouldSaveInstrument,
-                            confirm: false,
-                        },
-                        shouldSetAsDefaultInstrument,
-                    },
-                };
-
-                return await this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload));
-            }
-
-            const { paymentIntent, error } = await this._stripeUPEClient.confirmPayment({
-                elements: this._stripeElements,
-                redirect: 'if_required',
-            });
+        if (isVaultedInstrument(paymentData)) {
+            await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
+            const { instrumentId } = paymentData;
 
-            if (error || !paymentIntent) {
-                if (error && includes(['card_error', 'invalid_request_error', 'validation_error'], error.type)) {
-                    throw new Error(error.message);
-                }
-                throw new RequestError();
-            }
+            return await this._executeWithVaulted(payment, instrumentId, shouldSetAsDefaultInstrument);
+        }
 
+        if (includes(APM_REDIRECT, methodId)) {
             await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
```

Comment:
Maybe a good idea to move this to it's own method, similar as `_executeWithVaulted` ?